src/channel.cr

```diff
@@ -286,42 +287,72 @@ class Channel(T)
       .uniq(&.lock_object_id)
       .sort_by(&.lock_object_id)
 
-    ops_locks.each &.lock
+    while true
+      ops_locks.each &.lock
 
-    ops.each_with_index do |op, index|
-      ignore = false
-      result = op.execute
+      ops.each_with_index do |op, index|
+        result = op.execute
 
-      unless result.is_a?(NotReady)
+        unless result.is_a?(NotReady)
+          ops_locks.each &.unlock
+          return index, result
+        end
+      end
+
+      if has_else
         ops_locks.each &.unlock
-        return index, result
+        return ops.size, NotReady
       end
-    end
 
-    if has_else
-      ops_locks.each &.unlock
-      return ops.size, NotReady
-    end
+      state = Atomic(SelectState).new(SelectState::Active)
+      all_closed = true
+      contexts = ops.map_with_index do |op, index|
+        # Avoid waiting over a closed channel
+        if op.closed?
+          nil
+        else
+          all_closed = false
+          op.create_context_and_wait(pointerof(state))
+          # Closing one of the channels will wakeup this path
+        end
+      end
 
-    state = Atomic(SelectState).new(SelectState::Active)
-    contexts = ops.map &.create_context_and_wait(pointerof(state))
+      if all_closed
+        ops_locks.each &.unlock
 
-    ops_locks.each &.unlock
-    Crystal::Scheduler.reschedule
+        # If channels are closed, there is no need to reschedule
+        # if there is no has_else an exception is more accurate probable
+        if has_else
+          return ops.size, NotReady
+        else
+          raise "All channels are closed on a blocking select"
```

Comment:
This should `raise ClosedError.new`, exactly as if you had used `receive` on any one of the closed channels.

*ideally* this would be designed to literally call `ch.receive` on the last channel to be closed, as that's semantically what's happening on my head, but that's not feasible.